### DISCUSSION TITLE
ExternalTaskSensor soft_fail=True does not work as expected

### DISCUSSION THREAD

**Discussion by romanzdk** [apache/airflow#31701] [Q&A] [answered]
### Apache Airflow version

Other Airflow 2 version (please specify below)

### What happened

I use ExternalTaskSensor to wait for another DAG, however I want the sensor to be marked as SKIPPED when the external DAG fails. I tried to add soft_fail=True but the sensor still failed.

The sensor:

```python
some_dag_sensor = ExternalTaskSensor(
	task_id = 'some_dag_sensor',
	external_dag_id = 'some-dag-0.0.1',
	execution_date_fn = lambda dt: (
		dt
		- datetime.timedelta(
			days = {calendar.MONDAY: 3, calendar.SUNDAY: 2, calendar.SATURDAY: 1}.get(dt.weekday(), 1)
		)
	),
	retries = 1,
	execution_timeout = datetime.timedelta(minutes = 15),
	check_existence = True,
	soft_fail=True, 
	dag = dag,
)
```

### What you think should happen instead

ExternalTaskSensor should be marked as SKIPPED when upstream task failed

### How to reproduce

Create ExternalTaskSensor task

```python
some_dag_sensor = ExternalTaskSensor(
	task_id = 'some_dag_sensor',
	external_dag_id = 'some-dag-0.0.1',
	execution_date_fn = lambda dt: (
		dt
		- datetime.timedelta(
			days = {calendar.MONDAY: 3, calendar.SUNDAY: 2, calendar.SATURDAY: 1}.get(dt.weekday(), 1)
		)
	),
	retries = 1,
	execution_timeout = datetime.timedelta(minutes = 15),
	check_existence = True,
	soft_fail=True, 
	dag = dag,
)
```

### Operating System

debian

### Deployment

Other Docker-based deployment

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by tirkarthi** [comment]
Please specify the Airflow version used. This PR seems related

https://github.com/apache/airflow/pull/23647

> Prior to this change, the ExternalTaskSensor ignored the soft_fail setting, and always marked it as an error. With this change in place, the ExternalTaskSensor will instead mark the task as skipped, as desired.

**Comment by romanzdk** [comment]
Airflow 2.5.2. It sounds the issue you sent should hopefully resolve it.

**Comment by phanikumv** [comment]
@romanzdk This issue is fixed in 2.3.3 and you are on 2.5.2, then why are you still getting this issue?

**Comment by romanzdk** [comment]
Well, good question. I do not know. Maybe it is required to explicitly also set "failed_states"?

**Comment by phanikumv** [comment]
As per docs -> `Using ``skipped_states`` allows the sensor to skip if the target fails`
Can you please try it?

**Comment by hussein-awala** [comment, marked as answer]
> Well, good question. I do not know. Maybe it is required to explicitly also set "failed_states"?

By default, the `failed_states` is None (empty list), an the `soft_fail` works only if the external task has a state listed in this list, so yes, you need to add failed explicitly to this list, an alternative solution is add failed state to `skipped_states` parameter.
